r2i/tflite/engine.cc

```diff
@@ -193,41 +198,120 @@ std::shared_ptr<r2i::IPrediction> Engine::Predict (std::shared_ptr<r2i::IFrame>
+  int output_size;
+  auto tensor_data = this->GetOutputTensorData(this->interpreter.get(),
+                     output_size, error);
+  if (r2i::RuntimeError::EOK != error.GetCode()) {
+    return nullptr;
+  }
+
   prediction->SetTensorValues(tensor_data, output_size);
 
   return prediction;
 }
 
 Engine::~Engine () {
   this->Stop();
+  this->interpreter.reset();
+}
+
+void Engine::SetupResolver(::tflite::ops::builtin::BuiltinOpResolver
+                           &/*resolver*/) {
+  // No implementation for tflite engine
+}
+
+void Engine::SetInterpreterContext(::tflite::Interpreter */*interpreter*/) {
+  // No implementation for tflite engine
+}
+
+void Engine::PreprocessInputData(const float *input_data, const int size,
+                                 ::tflite::Interpreter *interpreter, r2i::RuntimeError &error) {
+  const auto &input_indices = interpreter->inputs();
+  const auto *tensor = interpreter->tensor(input_indices[0]);
+
+  if (!input_data) {
+    error.Set (RuntimeError::Code::WRONG_API_USAGE, "Failed to get image data");
+    return;
+  }
+
+  if (kTfLiteUInt8 == tensor->type) {
+    auto *input_fixed_tensor = interpreter->typed_input_tensor<uint8_t>(0);
+
+    std::vector<uint8_t> input_data_fixed;
+    input_data_fixed.resize(size);
+    for (int index = 0; index < size; index++) {
+      input_data_fixed[index] = (uint8_t)input_data[index];
+    }
+
+    memcpy(input_fixed_tensor, input_data_fixed.data(), input_data_fixed.size());
+  } else if (kTfLiteFloat32 == tensor->type) {
+    auto input_tensor = interpreter->typed_tensor<float>(input_indices[0]);
+
+    memcpy(input_tensor, input_data, size * sizeof(float));
+  } else {
+    error.Set (RuntimeError::Code::WRONG_API_USAGE,
+               "Output tensor has unsupported output type");
+    return;
+  }
+}
+
+float *Engine::GetOutputTensorData(::tflite::Interpreter *interpreter,
+                                   int &output_size,
+                                   r2i::RuntimeError &error) {
+  std::vector<float> output_data;
+  const auto &output_indices = interpreter->outputs();
+  const int num_outputs = output_indices.size();
+
+  int out_idx = 0;
+  for (int index = 0; index < num_outputs; ++index) {
+    const auto *out_tensor = interpreter->tensor(output_indices[index]);
+
+    if (nullptr == out_tensor) {
+      error.Set (RuntimeError::Code::FRAMEWORK_ERROR,
+                 "Output tensor is null");
+      return nullptr;
+    }
+
+    if (kTfLiteUInt8 == out_tensor->type) {
+
+      const int num_values = out_tensor->bytes;
+      output_data.resize(out_idx + num_values);
+      const uint8_t *output = interpreter->typed_output_tensor<uint8_t>(index);
+
+      for (int value_index = 0; value_index < num_values; ++value_index) {
+        output_data[out_idx++] = (output[value_index] - out_tensor->params.zero_point) *
+                                 out_tensor->params.scale;
+      }
```

Comment:
This smells like we can refactor into a templated function and recycle for several fixed point types